R/NCBI_snp_query.R

```diff
@@ -130,65 +134,78 @@ ncbi_snp_query <- function(SNPs, key = NULL, ...) {
     }
-    my_snpClass <- tryget(my_list$.attrs["snpClass"])
-
-    my_gene <- xml2::xml_attr(
-      xml2::xml_find_first(x2kids[[i]], "Assembly//Component/MapLoc/FxnSet"),
-      "symbol"
+    my_snpClass <- tryget(my_list$SNP_CLASS)
+    #my_fxnClass <- tryget(my_list$FXN_CLASS)
+    
+   # my_gene <- xml2::xml_attr(
+  #    xml2::xml_find_first(x2kids[[i]], "Assembly//Component/MapLoc/FxnSet"),
+   #   "symbol"
+    #)
+    my_gene <- tryget(
+      sapply(my_list$GENES, function(x) x$NAME)
     )
-    # my_gene <- tryget( my_list$Assembly$Component$MapLoc$FxnSet['symbol'] )
     if (is.null(my_gene)) my_gene <- NA
-    alleles <- my_list$Ss$Sequence$Observed
-
+    
+    
+    # don't really know whats in $SS, but $DOCSUM contains the alleles  
+    # #ALLELE contains...
+    # alleles <- my_list$SS$Sequence$Observed
+    meta_info_ <- tryget(my_list$DOCSUM) 
+    ## pull out what is after SEQ and remove SEQ=[ and ]
+    meta_info <- strsplit(meta_info_, "|", fixed=TRUE)[[1]][2]
+    alleles_ordered <- gsub("\\]", "", gsub("SEQ=\\[", "", meta_info))
+    
+    alleles <- strsplit(alleles_ordered, "/")[[1]]
+    ancestral_allele <- alleles[1]
+    variation_allele <- paste(alleles[-1], collapse = ",")
+    ## todo, cases of more than one alternate allele
+    
     ## handle true SNPs
-    if (my_snpClass %in% c("snp", "snv")) {
-      tmp <- c( my_list$Ss$Sequence$Observed, my_list$Ss$.attrs["orient"] )
-      if (tmp[2] != "forward") {
-        tmp[1] <- flip( tmp[1], sep = "/" )
-      }
-      alleles_split <- strsplit( tmp[1], "/" )[[1]]
-
-      ## check which of the two alleles grabbed is actually the minor allele
-      ## we might have to 'flip' the minor allele if there is no match
-      maf_allele <- my_list$Frequency['allele']
-      if (is.null(maf_allele)) {
-        maf_allele <- alleles <- strsplit(my_list$Sequence$Observed, "/")[[1]]
-        my_major <- alleles[1]
-        my_minor <- alleles[2]
-        my_freq <- NA
-      } else {
-        my_minor <- alleles_split[ maf_allele == alleles_split ]
-        my_major <- alleles_split[ maf_allele != alleles_split ]
-        my_freq <- my_list$Frequency["freq"]
-      }
+    if (my_snpClass %in% c("snp", "snv", "delins", "indel", "del") & any(my_list$GLOBAL_MAFS != "\n\t")) {
+   
+      ## pull out MAF, and the allele that its computed for. 
+      maf_df_ <- do.call("rbind", lapply(my_list$GLOBAL_MAFS, function(x) cbind(x$STUDY, x$FREQ)))
+      maf_df_2 <- do.call("rbind", strsplit(maf_df_[,2], "=|/"))
+      maf_df <- data.frame(cbind(maf_df_, maf_df_2))
+      names(maf_df) <- c("source", "info", "allele", "freq", "n")
+      maf_df$source <- as.character( maf_df$source)
+      maf_df$info <- as.character( maf_df$info)
+      maf_df$allele <- as.character(maf_df$allele)
+      maf_df$freq <- as.numeric(as.character(maf_df$freq))
+      
+      ## for specific source
+      maf_df_sub <- maf_df[maf_df$source == "GnomAD", ]
+      my_minor <- maf_df_sub$allele
```

Comment:
Hi @sinarueeger,
This is a followup from my comment regarding the minor allele reported when there is a duplication. From the tests there is rs146107628 which is a duplication of a T where the minor allele is reported as "T", whereas I think it should be "TT" (but perhaps I am wrong here?). It seems that in the esummary this arises because for these frequencies they are reporting the inserted sequence, which is "T" (and not the full variant which is "TT"), but this is perhaps a bit confusing (or at least to me). What do you think?

Reply:
I agree - this should indeed be "TT". 
The weird thing is, that although the browser reports `dupT=0.018` for 1000Genomes, the xml contains
```
$GLOBAL_MAFS$MAF$STUDY
[1] "1000Genomes"

$GLOBAL_MAFS$MAF$FREQ
[1] "T=0.0182/91"

```

Reply:
Hi @jooolia, 
One other (related) confusing thing is, that for example for https://www.ncbi.nlm.nih.gov/snp/rs1173690113, they report two alleles: `ACAGTCAGTATCAATTCTGGAAGAATTTCCAGACA` and `ACA`. The latter is a deletion of `GTCAGTATCAATTCTGGAAGAATTTCCAGACA`. But in the MAF list, they only report `-` (as a symbol for deletion). 
Not only is the deletion not indicated, but I can also not recognize from the xml output, which part is deleted (`GTCAGTATCAATTCTGGAAGAATTTCCAGACA` or `ACAGTCAGTATCAATTCTGGAAGAATTTCCAG`). Not sure we need to report this to that detail though.

Reply:
Hi @sinarueeger, hmm yes this esummary format for indels appears to have some issues. What to do? 1) If record type not "snv" use the variation allele for the minor allele, 2) use as reported by NCBI for now, but include warning in the documentation 3) something else?

Reply:
Hi @sinarueeger,
I emailed NCBI regarding this issue with the allele reported with the MAF issue with Esummary, however this seems to be not an issue with the results from the API.

Reply:
Thanks for emailing them. Fingers crossed!

Reply:
Hi @sinarueeger,
Still thinking about this more. How I understand it is that the issue here is the representation of the variant and whether they use an anchor or not. 
For this example (and super useful to have all of these tests), the alleles are "ACAGTCAGTATCAATTCTGGAAGAATTTCCAGACA, ACA" where there is some ambiguity. In the frequency data the deleted sequence is reported as "ACAGTCAGTATCAATTCTGGAAGAATTTCCAG" and the inserted as "-" thus giving a delACAGTCAGTATCAATTCTGGAAGAATTTCCAG. Could the minor allele for this frequency be reported as delACAGTCAGTATCAATTCTGGAAGAATTTCCAG? I will keep thinking about this, but I thought you might have some insights.  :) The data look similar in the new api format

Reply:
Great to hear that the data looks similar in the new api format!